### PR TITLE
Fix projection with optimize_aggregators_of_group_by_keys

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
1. Fix incorrect data of some aggregate projections when `optimize_aggregators_of_group_by_keys = true`. This fixes 
https://github.com/ClickHouse/ClickHouse/issues/49695

2. Fix incorrect data of some aggregate projections when `optimize_group_by_function_keys = true`. This fixes 
https://github.com/ClickHouse/ClickHouse/issues/40514

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Parsers/ASTProjectionSelectQuery.cpp [hunk 1]
```diff
@@ -142,6 +142,13 @@ ASTPtr ASTProjectionSelectQuery::cloneToASTSelect() const
     }
     if (groupBy())
         select_query->setExpression(ASTSelectQuery::Expression::GROUP_BY, groupBy()->clone());
+
+    auto settings_query = std::make_shared<ASTSetQuery>();
+    SettingsChanges settings_changes;
+    settings_changes.insertSetting("optimize_aggregators_of_group_by_keys", false);
+    settings_query->changes = std::move(settings_changes);
+    settings_query->is_standalone = false;
+    select_query->setExpression(ASTSelectQuery::Expression::SETTINGS, std::move(settings_query));
```
Comment (nickitat) [inline src/Parsers/ASTProjectionSelectQuery.cpp:151]: is it possible that another `settings` were already set in this query?

Comment (amosbird) [reply src/Parsers/ASTProjectionSelectQuery.cpp:151]: Nope. ProjectionSelectQuery doesn't allow setting clause

**Review by nickitat — APPROVED** [review]

**Comment by amosbird** [comment]
@nickitat I think it's OK to merge.

**Comment by nickitat** [comment]
integration tests failures look suspicious
https://play.clickhouse.com/play?user=play#c2VsZWN0IAp0b1N0YXJ0T2ZIb3VyKGNoZWNrX3N0YXJ0X3RpbWUpIGFzIGQsCmNvdW50KCksIGdyb3VwVW5pcUFycmF5KHB1bGxfcmVxdWVzdF9udW1iZXIpLCAgYW55KHJlcG9ydF91cmwpCmZyb20gY2hlY2tzIHdoZXJlICcyMDIzLTAxLTAxJyA8PSBjaGVja19zdGFydF90aW1lIGFuZCB0ZXN0X25hbWUgbGlrZSAnJXRlc3RfY2x1c3Rlcl9jb3BpZXIlJyBhbmQgdGVzdF9zdGF0dXMgaW4gKCdGQUlMJywgJ0ZMQUtZJykgZ3JvdXAgYnkgZCBvcmRlciBieSBkIGRlc2M=

**Comment by amosbird** [comment]
https://github.com/ClickHouse/ClickHouse/issues/30399   It seems related.

**Comment by amosbird** [comment]
It seems this PR can break `test_cluster_copier` integration tests in non-obvious way. I spent quite some efforts on it but cannot run these tests locally, thus cannot reproduce it yet. @nickitat Could you help take another look?

**Comment by amosbird** [comment]
@nickitat Thanks for the tips. I've added another fix to bypass setting constraints for projection ASTs. I think it should be fine now.

**Comment by amosbird** [comment]
@nickitat All tests have passed. Please take another look

**Pull request merged**